### PR TITLE
Demo GCM integration

### PR DESCRIPTION
**Pull request by Sergio-Mira**
https://swrvedev.jira.com/browse/SWRVE-11823
Example of how to properly integrate push notifications in the demo app.
Uses new method to intercept Swrve pushes when having multiple providers.
Shiny new icons too!

### CODE REVIEW

**Comment by Sergio-Mira** [comment]
@dominicmarmionswrve Could you have a look?

File: SwrveSDKDemo/src/main/java/com/swrve/sdk/demo/CustomGcmReceiver.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    public void onReceive(Context context, Intent intent) {
+        boolean interceptedIntent = false;
+        // Call the Swrve intent service if the push contains the Swrve payload _p
+        if("com.google.android.c2dm.intent.RECEIVE".equals(intent.getAction())) {
+            Bundle extras = intent.getExtras();
+            if (extras != null) {
+                Object rawId = extras.get("_p");
+                String msgId = (rawId != null) ? rawId.toString() : null;
+                if (!SwrveHelper.isNullOrEmpty(msgId)) {
+                    // It is a Swrve push!
```
Comment (dominicmarmionswrve) [inline SwrveSDKDemo/src/main/java/com/swrve/sdk/demo/CustomGcmReceiver.java:30]: Can we add a SwrveLogger here?

File: SwrveSDKDemo/src/main/java/com/swrve/sdk/demo/CustomGcmReceiver.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+public class CustomGcmReceiver extends GcmReceiver {
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        boolean interceptedIntent = false;
+        // Call the Swrve intent service if the push contains the Swrve payload _p
+        if("com.google.android.c2dm.intent.RECEIVE".equals(intent.getAction())) {
+            Bundle extras = intent.getExtras();
+            if (extras != null) {
+                Object rawId = extras.get("_p");
```
Comment (dominicmarmionswrve) [inline SwrveSDKDemo/src/main/java/com/swrve/sdk/demo/CustomGcmReceiver.java]: We should put "_p" into a constant. Along with "com.google.android.c2dm.intent.RECEIVE". I'd suggest creating a new GCMConstants file?

File: SwrveSDKDemo/src/main/java/com/swrve/sdk/demo/CustomGcmReceiver.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+                String msgId = (rawId != null) ? rawId.toString() : null;
+                if (!SwrveHelper.isNullOrEmpty(msgId)) {
+                    // It is a Swrve push!
+                    interceptedIntent = true;
+                    ComponentName comp = new ComponentName(context.getPackageName(), "com.swrve.sdk.gcm.SwrveGcmIntentService");
+                    intent = intent.setComponent(comp);
+                    context.startService(intent);
```
Comment (dominicmarmionswrve) [inline SwrveSDKDemo/src/main/java/com/swrve/sdk/demo/CustomGcmReceiver.java:35]: Perhaps make a helper method in SwrveGcmIntentService to encapsulate this? 
eg: 
SwrveGcmIntentService.start(Context context)

Comment (dominicmarmionswrve) [inline SwrveSDKDemo/src/main/java/com/swrve/sdk/demo/CustomGcmReceiver.java:35]: or perhaps:
SwrveGcmIntentService.handleSwrveGcmPush(Context context)....or something to that effect

File: SwrveSDKDemo/src/main/java/com/swrve/sdk/demo/DemoApplication.java [hunk 4]
```diff
@@ -13,7 +16,24 @@
...
-            SwrveSDK.createInstance(this, YOUR_APP_ID, YOUR_API_KEY);
+            SwrveConfig config = new SwrveConfig();
+            // Comment the following line if you do not want to do GCM push notifications
+            config.setSenderId(YOUR_SENDER_ID);
+
+            SwrveSDK.createInstance(this, YOUR_APP_ID, YOUR_API_KEY, config);
+
+            // Comment the following lines if you do not want to do GCM push notifications
+            // React to the push notification when the user clicks on it
+            SwrveSDK.setPushNotificationListener(new ISwrvePushNotificationListener() {
```
Comment (dominicmarmionswrve) [inline SwrveSDKDemo/src/main/java/com/swrve/sdk/demo/DemoApplication.java:27]: Is it standard behaviour for users to use the listener? If not, then I'd comment these lines out by default

**Comment by dominicmarmionswrve** [comment]
Looks good! I would prefer to see this as a separate sample app which focuses on push integration. Something we can do later.

**Comment by dominicmarmionswrve** [comment]
Also, this needs a jira, and properly dev tested.

File: SwrveSDKDemo/src/main/java/com/swrve/sdk/demo/DemoApplication.java (hunk 4 above)
Comment (Sergio-Mira) [inline SwrveSDKDemo/src/main/java/com/swrve/sdk/demo/DemoApplication.java:27]: They usually want to react to the push payload to move to a different part of the app etc

Comment (dominicmarmionswrve) [inline SwrveSDKDemo/src/main/java/com/swrve/sdk/demo/DemoApplication.java:27]: :+1:

File: SwrveSDK/src/google/java/com/swrve/sdk/gcm/SwrveGcmConstants.java [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package com.swrve.sdk.gcm;
+
+public final class SwrveGcmConstants {
```
Comment (dominicmarmionswrve) [inline SwrveSDK/src/google/java/com/swrve/sdk/gcm/SwrveGcmConstants.java:3]: :+1:

File: SwrveSDKDemo/src/main/java/com/swrve/sdk/demo/CustomGcmReceiver.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        boolean interceptedIntent = false;
+        // Call the Swrve intent service if the push contains the Swrve identifier
+        if("com.google.android.c2dm.intent.RECEIVE".equals(intent.getAction())) {
+            Bundle extras = intent.getExtras();
+            if (extras != null) {
+                Object rawId = extras.get(SwrveGcmConstants.SWRVE_TRACKING_KEY);
+                String msgId = (rawId != null) ? rawId.toString() : null;
+                if (!SwrveHelper.isNullOrEmpty(msgId)) {
+                    // It is a Swrve push!
+                    SwrveLogger.d("Received Swrve push, starting " + SwrveGcmConstants.SWRVE_DEFAULT_INTENT_SERVICE + " instead");
```
Comment (dominicmarmionswrve) [inline SwrveSDKDemo/src/main/java/com/swrve/sdk/demo/CustomGcmReceiver.java:31]: :+1:

**Comment by dominicmarmionswrve** [comment]
:rocket:

**Pull request merged**